### PR TITLE
improve handling of large single-part s3 uploads

### PR DESCRIPTION
**Pull request by icewind1991**
- [use php://temp instead of php://memory for multi-part upload buffer](https://github.com/nextcloud/server/commit/573825cec2207e0b58ff0410b03cfb1adb88d8a8)
- [don't perform the extra buffering in s3 stream write when the stream size is known](https://github.com/nextcloud/server/commit/9b842fe4914bb281d0aa6f804cc9ea1ed375dc9e)
- [don't perform unneeded seeks in assemblystream](https://github.com/nextcloud/server/commit/e4d0882b312923ee9021312c5272df4884276961)
- [try to read the full requested size in assemblystream even if the chunks are smaller](https://github.com/nextcloud/server/commit/e4d0882b312923ee9021312c5272df4884276961)
- [use the size from the stream when writing objects when possible](https://github.com/nextcloud/server/commit/6cf66f95ce9dadffd9fa25203e737917cef0cd8c)

This removes the risk of the memory size ballooning in cases where `putSizeLimit` has been made very high (because the setup doesn't have a working multi-part upload for example)

### CODE REVIEW

**Review by provokateurin — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Review by artonge — APPROVED** [review]

**Comment by icewind1991** [comment]
Failing checks are relevant, still working on this

**Comment by icewind1991** [comment]
re-requesting reviews since there were some significant changes

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 1]
```diff
@@ -457,6 +457,14 @@ public function file_put_contents(string $path, mixed $data): int {
 	}
 
 	public function writeStream(string $path, $stream, ?int $size = null): int {
+		if ($size === null) {
+			$stats = fstat($stream);
+			if (is_array($stats) && isset($stats['size'])) {
+				$size = $stats['size'];
+				$this->logger->warning("stream size $size");
```
Comment (juliusknorr) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:464]: debugging leftover?

**Review by juliusknorr — APPROVED** [review]
small comment, otherwise still 👍

**Review by provokateurin — APPROVED** [review]

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php (hunk 1 above)
Comment (icewind1991) [reply lib/private/Files/ObjectStore/ObjectStoreStorage.php:464]: https://github.com/nextcloud/server/pull/49739

**Comment by marcelklehr** [comment]
Can we backport this to stable30?

**Pull request merged**